cmd/fund/fund.go

```diff
@@ -1,312 +1,262 @@
-		return 0, jsonErr
+	// Deploy or instantiate the Funder contract.
+	var contract *funder.Funder
+	contract, err = deployOrInstantiateFunderContract(ctx, c, tops, privateKey)
+	if err != nil {
+		return err
 	}
 
-	// Extract the chain ID from the response
-	chainIDHex, ok := result["result"].(string)
-	if !ok {
-		return 0, fmt.Errorf("unable to extract chain ID from response")
+	// Derive or generate a set of wallets.
+	var addresses []common.Address
+	if params.WalletAddresses != nil && *params.WalletAddresses != nil {
+		log.Info().Msg("Using addresses provided by the user")
+		addresses = make([]common.Address, len(*params.WalletAddresses))
+		for i, address := range *params.WalletAddresses {
+			addresses[i] = common.HexToAddress(address)
+		}
+	} else if *params.UseHDDerivation {
+		log.Info().Msg("Deriving wallets from the default mnemonic")
+		addresses, err = deriveHDWallets(int(*params.WalletsNumber))
+	} else {
+		log.Info().Msg("Generating random wallets")
+		addresses, err = generateWallets(int(*params.WalletsNumber))
+	}
+	if err != nil {
+		return err
 	}
 
-	// Convert the chain ID from hex to int64
-	int64ChainID, parseErr := strconv.ParseInt(chainIDHex, 0, 64)
-	if parseErr != nil {
-		return 0, parseErr
+	// Fund wallets.
+	if err = fundWallets(ctx, c, tops, contract, addresses); err != nil {
+		return err
 	}
+	log.Info().Msg("Wallet(s) funded! 💸")
 
-	return int64ChainID, nil
+	log.Info().Msgf("Total execution time: %s", time.Since(startTime))
+	return nil
 }
 
-func generateNonce(web3Client *web3.Web3) (uint64, error) {
-	nonceMutex.Lock()
-	defer nonceMutex.Unlock()
-
-	if nonceInitialized {
-		globalNonce++
-	} else {
-		// Derive the public key from the funding wallet's private key
-		fundingWalletECDSA, ecdsaErr := crypto.HexToECDSA(fundingWalletPK)
-		if ecdsaErr != nil {
-			log.Error().Err(ecdsaErr).Msg("Error getting ECDSA from funding wallet private key")
-			return 0, ecdsaErr
-		}
+// dialRpc dials the Ethereum RPC server and return an Ethereum client.
+func dialRpc(ctx context.Context) (*ethclient.Client, error) {
+	rpc, err := rpc.DialContext(ctx, *params.RpcUrl)
+	if err != nil {
+		log.Error().Err(err).Msg("Unable to dial")
+		return nil, err
+	}
+	defer rpc.Close()
 
-		fundingWalletPublicKey = &fundingWalletECDSA.PublicKey
-		// Convert ecdsa.PublicKey to common.Address
-		fundingAddress := crypto.PubkeyToAddress(*fundingWalletPublicKey)
+	rpc.SetHeader("Accept-Encoding", "identity")
+	return ethclient.NewClient(rpc), nil
+}
 
-		nonce, err := web3Client.Eth.GetNonce(fundingAddress, nil)
-		if err != nil {
-			log.Error().Err(err).Msg("Error getting nonce")
-			return 0, err
-		}
-		globalNonce = nonce
-		nonceInitialized = true
+// Initialize  parameters.
+func initializeParams(ctx context.Context, c *ethclient.Client) (*ecdsa.PrivateKey, *big.Int, error) {
+	// Parse the private key.
+	privateKey, err := crypto.HexToECDSA(strings.TrimPrefix(*params.PrivateKey, "0x"))
```

Comment:
nice, just added this to `polycli loadtest` too. given that the `--private-key` flag is used in a lot of place, the logic of trimming the `0x` will be reused a lot and we may have to eventually think of a approach to streamline in (maybe putting it in the `util` package, idk yet)